django/core/serializers/python.py

```diff
@@ -52,43 +56,42 @@ def handle_field(self, obj, field):
         self._current[field.name] = self._value_from_field(obj, field)
 
     def handle_fk_field(self, obj, field):
-        if self.use_natural_foreign_keys and hasattr(
-            field.remote_field.model, "natural_key"
-        ):
-            related = getattr(obj, field.name)
-            if related:
-                value = related.natural_key()
+        if self.use_natural_foreign_keys:
+            natural_key_value = self._resolve_fk_natural_key(obj, field)
+            if natural_key_value is not None:
+                value = natural_key_value
             else:
-                value = None
+                value = self._value_from_field(obj, field)
         else:
             value = self._value_from_field(obj, field)
         self._current[field.name] = value
 
     def handle_m2m_field(self, obj, field):
         if field.remote_field.through._meta.auto_created:
-            if self.use_natural_foreign_keys and hasattr(
-                field.remote_field.model, "natural_key"
-            ):
+            use_natural = (
+                self.use_natural_foreign_keys
+                and self._model_supports_natural_key(field.remote_field.model)
+            )
```

Comment:
This refactor isn't strictly necessary, so reverting it will help clarify your (understandably) large diff.